File: packages/sdk-php/src/ConnectorsAPI.php
<?php

declare(strict_types=1);

namespace Melaya;

/**
 * Project Connectors API — manage credentials at project scope.
 *
 * Maps to /api/v1/private/projects/:project/connectors/*. Project-scoped
 * connectors are isolated per project, letting separate projects use different
 * API keys for the same service.
 *
 * @example
 * ```php
 * $sdk = new Melaya\Melaya(apiKey: getenv('MK'));
 *
 * $sdk->connectors->set('my-project', 'openai', ['value' => 'sk-...']);
 * $services = $sdk->connectors->connectedServices('my-project');
 *
 * $handle = $sdk->connectors->envHandle('my-project');
 * // $handle['token'] is used by the runner to decrypt credentials
 * ```
 */
class ConnectorsAPI
{
    public function __construct(private readonly HttpClient $http) {}

    /** List connected services for a project. */
    public function connectedServices(string $project): array
    {
        return $this->http->get(
            '/api/v1/private/projects/' . rawurlencode($project) . '/connectors/services'
        );
    }

    /**
     * Store a connector credential at project scope.
     *
     * @param string $project Project name
     * @param string $service Service name (e.g. 'openai')
     * @param array  $body    ['value' => '...', 'key' => '...', 'label' => '...']
     */
    public function set(string $project, string $service, array $body): array
    {
        return $this->http->put(
            '/api/v1/private/projects/' . rawurlencode($project) . '/connectors/' . rawurlencode($service),
            $body
        );
    }

    /** Delete a project-scoped connector credential. */
    public function delete(string $project, string $service): array
    {
        return $this->http->delete(
            '/api/v1/private/projects/' . rawurlencode($project) . '/connectors/' . rawurlencode($service)
        );
    }

    /**
     * Get a short-lived env-handle token for project-scoped credentials.
     * The runner uses this token to decrypt credentials without a full session.
     */
    public function envHandle(string $project): array
    {
        return $this->http->post(
            '/api/v1/private/projects/' . rawurlencode($project) . '/connectors/env-handle'
        );
    }

    /**
     * Start Google OAuth flow for a project-scoped connector.
     *
     * @param string $project Project name
     * @param array  $body    OAuth config body
     */
    public function googleOAuthStart(string $project, array $body = []): array
    {
        return $this->http->post(
            '/api/v1/private/projects/' . rawurlencode($project) . '/connectors/google/oauth',
            $body ?: null
        );
    }
}

File: packages/sdk-java/src/main/java/org/melaya/ConnectorsAPI.java
package org.melaya;

import com.fasterxml.jackson.databind.JsonNode;

import java.util.Map;

/**
 * Project Connectors API — manage credentials at project scope.
 *
 * <p>Maps to {@code /api/v1/private/projects/:project/connectors/*}.
 * Project-scoped connectors are isolated per project, allowing different
 * projects to use different API keys for the same service.
 *
 * <p>For user-scoped credentials see {@link CredentialsAPI}.
 *
 * @example
 * <pre>{@code
 * melaya.connectors().set("my-project", "openai", Map.of("value", "sk-..."));
 * JsonNode services = melaya.connectors().connectedServices("my-project");
 * }</pre>
 */
public class ConnectorsAPI {

    private final HttpClient http;

    ConnectorsAPI(HttpClient http) {
        this.http = http;
    }

    /**
     * List connected services for a project.
     *
     * @param project the project name
     */
    public JsonNode connectedServices(String project) {
        return http.get("/api/v1/private/projects/" + encode(project) + "/connectors/services", null);
    }

    /**
     * Store or update a connector credential at project scope.
     *
     * @param project the project name
     * @param service the service/provider name
     * @param body    map containing at minimum {@code value}; optionally {@code key}, {@code label}
     */
    public JsonNode set(String project, String service, Map<String, Object> body) {
        return http.put(
                "/api/v1/private/projects/" + encode(project) + "/connectors/" + encode(service),
                body);
    }

    /**
     * Delete a project-scoped connector credential.
     *
     * @param project the project name
     * @param service the service name to remove
     */
    public JsonNode delete(String project, String service) {
        return http.delete(
                "/api/v1/private/projects/" + encode(project) + "/connectors/" + encode(service),
                null);
    }

    /**
     * Get a short-lived env-handle token for project-scoped credentials.
     * The runner uses this token to decrypt credentials without a full session.
     *
     * @param project the project name
     */
    public JsonNode getEnvHandle(String project) {
        return http.post(
                "/api/v1/private/projects/" + encode(project) + "/connectors/env-handle",
                null);
    }

    /**
     * Start a Google OAuth flow for a project-scoped connector.
     *
     * @param project the project name
     * @param body    OAuth flow config
     */
    public JsonNode googleOAuthStart(String project, Map<String, Object> body) {
        return http.post(
                "/api/v1/private/projects/" + encode(project) + "/connectors/google/oauth",
                body);
    }

    // ── helper ────────────────────────────────────────────────────────────────

    private static String encode(String s) {
        try {
            return java.net.URLEncoder.encode(s, java.nio.charset.StandardCharsets.UTF_8);
        } catch (Exception e) {
            return s;
        }
    }
}
